torchrec_dlrm/dlrm_main.py

```diff
@@ -499,75 +406,54 @@ class TrainValTestResults:
 
 def train_val_test(
     args: argparse.Namespace,
-    train_pipeline: TrainPipelineSparseDist,
+    pipeline: TrainPipelineSparseDist,
     train_dataloader: DataLoader,
     val_dataloader: DataLoader,
     test_dataloader: DataLoader,
-    lr_scheduler,
+    lr_scheduler: LRPolicyScheduler,
 ) -> TrainValTestResults:
     """
-    Train/validation/test loop. Contains customized logic to ensure each dataloader's
-    batches are used for the correct designated purpose (train, val, test). This logic
-    is necessary because TrainPipelineSparseDist buffers batches internally (so we
-    avoid batches designated for one purpose like training getting buffered and used for
-    another purpose like validation).
+    Train/validation/test loop.
 
     Args:
         args (argparse.Namespace): parsed command line args.
-        train_pipeline (TrainPipelineSparseDist): pipelined model.
-        train_dataloader (DataLoader): DataLoader used for training.
-        val_dataloader (DataLoader): DataLoader used for validation.
-        test_dataloader (DataLoader): DataLoader used for testing.
+        pipeline (TrainPipelineSparseDist): pipelined model.
+        train_dataloader (DataLoader): Training set's dataloader.
+        val_dataloader (DataLoader): Validation set's dataloader.
+        test_dataloader (DataLoader): Test set's dataloader.
+        lr_scheduler (LRPolicyScheduler): Scheduler to adjust the learning rate.
 
     Returns:
         TrainValTestResults.
     """
 
     train_val_test_results = TrainValTestResults()
 
-    train_iterator = iter(train_dataloader)
-    test_iterator = iter(test_dataloader)
     for epoch in range(args.epochs):
-        val_iterator = iter(val_dataloader)
+        new_pipeline = TrainPipelineSparseDist(pipeline._model, pipeline._optimizer, pipeline._device)
```

Comment:
we use new_pipeline here, but never use pipeline passed in. why?

Reply:
The original idea was to treat `pipeline` as the source of the context object and have `pipeline` be in charge of setting the contexts of other pipelines created.

I have refactored the code more, and in the newer code I use `pipeline` as shown below. I refactored by creating the function:
```
def new_pipeline(pipeline: TrainPipelineSparseDist) -> TrainPipelineSparseDist:
    empty_pipeline = TrainPipelineSparseDist(pipeline._model, pipeline._optimizer, pipeline._device)
    empty_pipeline._context = pipeline._context
    return empty_pipeline
```
and using this function such as:
```
    for epoch in range(args.epochs):
        _train(
            new_pipeline(pipeline),
            train_dataloader,
            val_dataloader,
            epoch,
            args.batch_size,
            lr_scheduler,
            args.print_lr,
            args.validation_freq_within_epoch,
            args.limit_train_batches,
            args.limit_val_batches,
        )
        val_accuracy, val_auroc = _evaluate(args.limit_val_batches, new_pipeline(pipeline), val_dataloader, "val")
```

R